### DISCUSSION TITLE
Startup Extensibility

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#956] [General]
Problem:
- .NET Core applications have a single Startup class for registering and configuring services
- some services need to be declared in a specific order in order for them to work properly
- .NET Core has an IStartupFilter interface which allows developers to build a pipeline of Configure methods
- Oqtane has some mandatory services which are required in order for it to function correctly
- Oqtane has some services configured by default to enable functionality however it should be possible to override them with custom implementations
- third party developers need to be able to declare their own services during startup

Oqtane has IServerStartup and IClientStartup interfaces. They currently allow for multiple implementations within an installation. They are used to augment the default startup class with additional functionality. The problem is that there is no way to control the order of services, or replace any default services with alternate implementations.

Option 1

Use the current approach. Introduce config option to disable specific default implementations ( ie. DisableStartupServices="Authentication" ). The advantage to this approach is that it is simple and gives a developer control over disabling specific default services and replacing them. The disadvantage is that it does not support the ordering of services. 

Option 2

Only allow for a single active implementation of IServerStartup and IClientStartup interfaces in an installation. IServerStartup and IClientStartup would be expected to be a full plug-in replacement startup definition. An implementation would need to include all of the default Oqtane startup services ( we could simplify this by abstracting Oqtane startup logic into logical groups which could be more easily included ). The advantage to this approach is that it emulates a .NET Core native approach, uses the current interfaces, and gives a developer full control over ordering and adding/removing services. The disadvantage is that it changes the meaning of the current interfaces and is essentially a behaviorial breaking change ( not a binary breaking change ). 

I am sure there are many more options as well. Any suggestions would be helpful - especially from those of you who are already using the IServerStartup and IClientStartup interfaces in your installations.

**Comment by hishamco** [comment]
As I see option both options are not able to control the order of activating the core services, because some services are run before `IServerStartup` is called. I'm testing another option that I will share ASAP

**Reply by sbwalker** [reply]
Actually one of the main benefits of Option 2 above is that the developer has 100% control over service ordering. This is because they would be responsible for providing a full replacement for the startup methods - rather than just injecting additional services.

**Comment by hishamco** [comment]
I did an exterminate that can have the advantages of both options that you mentioned with order respect as follows:

- Introduce `IOqtaneServerStartupServices` & `IOqtaneClientStartupServices`, let us focus on server services for now

- `IOqtaneServerStartupServices` should have all the Oqtane services such as 'AddIdentity`, `AddLocalization`, `AddDatabase` .. etc

- To make it simpler for developer to extend the core services, `DefaultOqtaneServerServices` is introduces that mark all the services as `virtual` so the developers can pick the one who interested in to extend and reimplement and leave the others to the default implementation

- Oqtane Framework should be responsible to call the `IOqtaneServerStartupServices` core services in the right places without affecting the normal flow

- The important question is what if more than one module overriding the same services, IMHO the last should WIN like normal ASP.NET Core DI did

I can provide a PR as proof of concept and all the feedback are welcome

**Reply by sbwalker** [reply]
This sounds interesting but it does not describe how service order would be controlled. The disadvantage to the above approach is that it requires a whole new set of interfaces, and it would not be clear when developers should be using the new interfaces vs. the existing interfaces.

**Reply by hishamco** [reply]
> This sounds interesting but it does not describe how service order would be controlled

This will controlled by the Oqtane Framework itself, because it knows the right order and it will call the methods in the right time, I will create a draft PR to show you what I mean

> The disadvantage to the above approach is that it requires a whole new set of interfaces, and it would not be clear when developers should be using the new interfaces vs. the existing interfaces.

The `IOqtaneServerStartupServices` & `IOqtaneClientStartupServices` is a special version of `IServerIStartup` & `IClientStartup` and it should be used whenever the developer needs to override a certain implementation for any of Oqtane Framework core services, this should be documented too

The main issue with `IServerIStartup` & `IClientStartup` that they are run too late, so it's suited for registering developers services not Oqtane once

**Reply by sbwalker** [reply]
If we abandoned the idea that Oqtane itself needs to be registered using an IStartup interface and just used the current approach, then the only real problem which needs to be solved is how to disable the authentication service registration and allow a developer to replace it with an alternative, correct?

**Reply by hishamco** [reply]
Correct, to disable the authentication we could use features managements and for replacing the service we should add the services to DI with any of the suggested options, taking in mind WHEN should be registered

**Comment by sbwalker** [comment]
There are a number of general questions which need to be answered in regards to Startup Extensibility:

- who is this feature primarily focused at? I believe the current interface is being used mainly by people registering component libraries at the global level so that they can be used in all modules in a site. Is this the main use case? Or would module developers expect to use it and ship an implementation of this interface with their packaged modules?
- how many core services do we imagine can realistically be replaced with alternatives? So far the only core service which someone has requested the ability to override is authentication. Are there any others?
- is manipulating the ordering of services a common requirement or an exception case? No one has actually provided a practical use case yet where this is an issue.

**Comment by hishamco** [comment]
> who is this feature primarily focused at? I believe the current interface is being used mainly by people registering component libraries at the global level so that they can be used in all modules in a site. Is this the main use case? Or would module developers expect to use it and ship an implementation of this interface with their packaged modules?

AFAIK the aim of startup interfaces are to allow the module developers to inject their services into the pipeline, but sometime they may need to replace a default implementation for the core implementation, rather than shipping this in multiple modules, the framework could provide a way to override these core services even from multiple modules, say Module 1 will override logging and Module 2 overrides authentication

> how many core services do we imagine can realistically be replaced with alternatives? So far the only core service which someone has requested the ability to override is authentication. Are there any others?

- Authentication

- Localization

- Data Storage

- Logging

- File Storage


and much more ..


>is manipulating the ordering of services a common requirement or an exception case? No one has actually provided a practical use case yet where this is an issue.

The services order is a common requirement for both Oqtane Framework services and developer services, as you know changing the order of the startup services will lead to a lot of problematic in Oqtane pipeline, also there are use case when I as developer need to register a services into a specific order for example I may need to inject a health check services before that Oqtane is rendered in the browser or inject a services before or after localization. IMHO service ordering is very important but the order of the Oqtane core services should be handled by the framework itself to avoid pipeline issues

**Reply by sbwalker** [reply]
I don’t think your response answered my question... but it’s probably a matter of terminology. We need a better way to define the actors. I see an Application Developer as someone who is using Oqtane to build a custom web application. The application will include custom modules and themes for their own specific needs. A Module Developer is focused on building reusable modules which will be packaged and distributed to others for use in their applications. So I was curious which of these actors we would expect to use the IStartup interface. I was assuming the former rather than the latter based on what I have seen so far. But I could be mistaken.

**Reply by sbwalker** [reply]
Most of the services you identified above could actually not be replaced currently in Oqtane because they were not designed with replacement in mind. It would take more effort than just swapping a startup registration to replace their functionality. This is exactly why I asked the question above, as if people think a new startup interface may allow them to use a different database or a different logging library, etc... they would be mistaken.

**Reply by sbwalker** [reply]
I was hoping for an actual use case which required specific service ordering - not a theoretical or hypothetical case. I have learned the hard way over time that theoretical use cases rarely behave as you expect so it not a good idea to build solutions for them. Instead it is better to build solutions only once you have a few practical real world cases to solve.

**Reply by hishamco** [reply]
I don’t think your response answered my question

May be I miss understood you ;)

as if people think a new startup interface may allow them to use a different database or a different logging library, etc... they would be mistaken.

Naming is hard ;) please few minutes and I will refer you to my prototype, I think I made a mistake with interface naming that is why they are confusing

**Reply by hishamco** [reply]
> Most of the services you identified above could actually not be replaced currently in Oqtane because they were not designed with replacement in mind

It's not necessary to be replaced, the main aim is extensibility, so most of time they need to be extend rather than replacing, but in case of logging for instance may be the requirement is to use another logger type other that database

**Comment by sbwalker** [comment]
One other option to consider.... do nothing right now to try and solve this problem and wait until we properly tackle OAuth before introducing any long term solution which we need to support/maintain. Since this is open source, if a developer truly needs to customize startup they can modify the source in their local fork. The advantage is that we will most likely make wiser decisions once we have more use cases to consider. The disadvantage is that the upgrade path may be a bit more complicated for those who absolutely need to make these changes now. However they may also be able to contribute their knowledge and code back to the project - which is the whole idea behind open source.

**Comment by sbwalker** [comment]
@hishamco have you ever used the Replace() method on ServiceCollectionDescriptorExtensions:

https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.extensions.servicecollectiondescriptorextensions.replace?view=dotnet-plat-ext-5.0

ServiceCollection is just a wrapper on top of List&lt;ServiceDescriptor&gt; so this allows replacement of default services.

**Reply by hishamco** [reply]
Ya, this is another option that we could use to replace the already registered services, it is handy in some cases, but the problem we are facing here is that `IStartupServices` run at the late stage, which may cause some issues before my replace is done

**Reply by hishamco** [reply]
This is a use case where we use `Replace` in `OrchardCore.Localization` module 
https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Localization/Startup.cs#L32-L33
